backend/remote-state/kubernetes/client.go

```diff
@@ -0,0 +1,363 @@
+	if !ok {
+		// The secret exists but there is no state in it
+		return nil, nil
+	}
+
+	stateRawString := stateRaw.(string)
+
+	state, err := uncompressState(stateRawString)
+	if err != nil {
+		return nil, err
+	}
+
+	md5 := md5.Sum(state)
+
+	p := &remote.Payload{
+		Data: state,
+		MD5:  md5[:],
+	}
+	return p, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	secretName, err := c.createSecretName()
+	if err != nil {
+		return err
+	}
+
+	payload, err := compressState(data)
+	if err != nil {
+		return err
+	}
+
+	secret, err := c.getSecret(secretName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		return fmt.Errorf("secret does not exist so lock is not held, secret name: %v err: %v", secretName, err)
+	}
+
+	lockInfo, err := c.getLockInfo(secret)
+	if err != nil {
+		return err
+	}
+
+	setState(secret, payload)
+
+	secret, err = c.kubernetesSecretClient.Update(secret, metav1.UpdateOptions{})
+	if err != nil {
+		lockErr := &state.LockError{
+			Info: lockInfo,
+			Err:  fmt.Errorf("error updating the state: %v", err),
+		}
+		return lockErr
+	}
+
+	return err
+}
+
+// Delete the state secret
+func (c *RemoteClient) Delete() error {
+	secretName, err := c.createSecretName()
+	if err != nil {
+		return err
+	}
+
+	err = c.deleteSecret(secretName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	secretName, err := c.createSecretName()
+	if err != nil {
+		return "", err
+	}
+
+	lockInfo := info.Marshal()
+
+	secret, err := c.getSecret(secretName)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return "", err
+		}
+
+		// The secret doesn't exist yet, create it with the lock
+		secretData := make(map[string][]byte)
+		secretData[tfstateLockInfoKey] = lockInfo
+
+		labels := map[string]string{
+			tfstateKey:             "true",
+			tfstateSecretSuffixKey: c.nameSuffix,
+			tfstateWorkspaceKey:    c.workspace,
+			managedByKey:           "terraform",
+		}
+
+		if len(c.labels) != 0 {
+			for k, v := range c.labels {
+				labels[k] = v
+			}
+		}
+
+		secret = &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"metadata": metav1.ObjectMeta{
+					Name:        secretName,
+					Namespace:   c.namespace,
+					Labels:      labels,
+					Annotations: map[string]string{"encoding": "gzip"},
+				},
+				"data": secretData,
+			},
+		}
+
+		_, err = c.kubernetesSecretClient.Create(secret, metav1.CreateOptions{})
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return "", err
+			}
+			// The secret was created between the get and create, grab it and keep going
+			secret, err = c.getSecret(secretName)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			// No error on the create so the lock has been created successfully
+			return info.ID, nil
+		}
+	}
+
+	li, err := c.getLockInfo(secret)
+	if err != nil {
+		return "", err
+	}
+
+	if li != nil {
```

Comment:
@alexsomesan

This unfortunately does not provide a safe lock. The big clue is that the lock is handled in multiple operations, but "check _then_ create" (or the reverse) cannot work with multiple concurrent clients. Some form of a single test-and-set operation is needed to create a lock (we don't require blocking locks for mutual exclusion, so more advanced primitives are not required)

Multiple clients can reach this point with the expectation that their lock info is going to be saved. The following `Update` call will then update the secret for both with the last client's info being saved, but both clients assuming they have a lock.

@alexsomesan, the comments above lead me to believe that kubernetes secrets may not have the ability to use this storage as a lock, but I'm not familiar with the data model of the secrets storage to know for sure.

Reply:
@jbardin That is not how the k8s API works. If you hold an object, in this case a secret, which has a version ID, and then attempt to make an update to that secret, the k8s API will validate you have the most recent version of the secret and if you do not, reject your update request. The lock is being enforced by the API by  assuring the client is only able to make updates to the most recent object.

Reply:
Thanks @dramich, I'll look into that. Testing this has consistently resulted in multiple clients obtaining locks concurrently, so I'll have to investigate further.

Reply:
Just to briefly update on this; I realized the client-go exposes the same [locking mechanism used](https://godoc.org/k8s.io/client-go/tools/leaderelection/resourcelock) by kubernete's controller-manager to do leader election, so I'm reworking this PR to use that.